### ISSUE TITLE
Stack level too deep when mutating included hook

### ISSUE THREAD

**Issue by pawelpacana** [mbj/mutant#1054]
We have recently switched from running mutant on Ruby 2.6 to 2.7. On the code that was not changed, and the same mutant version there are different results:

```
bundle list |grep mutant                                                                                                                  
  * mutant (0.9.12)
  * mutant-license (0.1.0)
  * mutant-rspec (0.9.10)
```

### 2.6.6p146

```
Mutant environment:
Matcher:         #<Mutant::Matcher::Config match_expressions: [AggregateRoot*]>
Integration:     rspec
Jobs:            8
Includes:        ["lib"]
Requires:        ["aggregate_root"]
Subjects:        27
Total-Tests:     35
Selected-Tests:  31
Tests/Subject:   1.15 avg
Mutations:       985
Results:         985
Kills:           985
Alive:           0
Runtime:         18.61s
Killtime:        107.46s
Overhead:        -82.69%
Mutations/s:     52.94
Coverage:        100.00%
```

### 2.7.0p0

```
Mutant environment:
Matcher:         #<Mutant::Matcher::Config match_expressions: [AggregateRoot*]>
Integration:     rspec
Jobs:            8
Includes:        ["lib"]
Requires:        ["aggregate_root"]
Subjects:        27
Total-Tests:     35
Selected-Tests:  31
Tests/Subject:   1.15 avg
Mutations:       985
Results:         985
Kills:           981
Alive:           4
Runtime:         17.27s
Killtime:        84.41s
Overhead:        -79.55%
Mutations/s:     57.05
Coverage:        99.59%

```

One of the failing mutations:

```
evil:AggregateRoot.included:/Users/pawelpacana/Code/rails_event_store/aggregate_root/lib/aggregate_root.rb:86:d3c17
@@ -1,4 +1,4 @@
 def self.included(host_class)
-  host_class.include(with_default_apply_strategy)
+  host_class.include(self)
 end
-----------------------
Killfork exited nonzero. Its result (if any) was ignored.
Process status:
#<Process::Status: pid 78659 SIGILL (signal 4)>
Killing the mutation resulted in an integration error.
This is the case when the tests selected for the current mutation
did not produce a test result, but instead an exception was raised.

This may point to the following problems:
* Bug in mutant
* Bug in the ruby interpreter
* Bug in your test suite
* Bug in your test suite under concurrency

The following exception was raised:


#<ArgumentError: marshal data too short>
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/isolation/fork.rb:98:in `load'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/isolation/fork.rb:98:in `read_child_result'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/isolation/fork.rb:69:in `call'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/procto-0.0.3/lib/procto.rb:62:in `public_send'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/procto-0.0.3/lib/procto.rb:62:in `block in initialize'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/isolation/fork.rb:169:in `block (2 levels) in call'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/isolation/fork.rb:32:in `block in with'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/isolation/fork.rb:31:in `pipe'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/isolation/fork.rb:31:in `with'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/isolation/fork.rb:168:in `block in call'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/isolation/fork.rb:32:in `block in with'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/isolation/fork.rb:31:in `pipe'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/isolation/fork.rb:31:in `with'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/isolation/fork.rb:167:in `call'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/env.rb:130:in `run_mutation_tests'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/env.rb:51:in `kill'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/parallel/worker.rb:28:in `call'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/parallel/worker.rb:28:in `block in call'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/parallel/worker.rb:23:in `loop'
/Users/pawelpacana/.rubies/ruby-2.7.0/lib/ruby/gems/2.7.0/gems/mutant-0.9.12/lib/mutant/parallel/worker.rb:23:in `call'

-----------------------
```

This is the mutation that causes stack error (when mutated manually) on the simplified code:

```ruby
module Xyz
  def self.some_module
    Module.new do
      # ...
    end
  end

  def self.included(klass)
    # klass.include(some_module) # original include
    klass.include(self)          # mutation
  end
end

class Abc
  include Xyz
end
```

**Comment by mbj** [comment]
@pawelpacana I think this is closely related to this one https://github.com/mbj/mutant/issues/965.

TLDR: Mutant is one of the few mutation testing engines that differentiates between:

A) Mutation that caused a test failure
B) Mutation that caused abnormal behavior.

Where the latter is *NOT* counted as a kill.

To verify this is a class of B) can you hand edit in the mutation and run rspec? And post the output of rspec here? If rspec crashes "hard" and does not exist "normally nonzero while printing an exit status" this means its an instance of the referenced issue. If rspec exits normally its a problem in mutant.

**Comment by pawelpacana** [comment]
@mbj Complete output after hand editing mutation:

https://gist.githubusercontent.com/pawelpacana/b6ec1155048a20ad61cb90fbbdff7780/raw/46f8ec45830c24d68bf1ef464f22c852442f8662/gistfile1.txt

```
Finished in 0.00003 seconds (files took 0.6934 seconds to load)
0 examples, 0 failures, 6 errors occurred outside of examples

```
```
echo $status                                                                                                                    
1
```

**Comment by mbj** [comment]
> 6 errors occurred outside of examples

Thats it. Mutant currently only respects `failures` not "errors outside of examples". As rspec *still* produces a result, I'd argue mutant should call mutations that cause these kinds of "meta test failures" as covered. So its not exactly the case of #965 where rspec fails to produce a result.

That should be an easy change to `mutant-rspec`.

**Comment by mbj** [comment]
OT: was your sample on ruby-2.6 or ruby-2.7, does it change between them? If so rspec authors probably want to know ;)

**Comment by pawelpacana** [comment]
Hand mutating leads to the same result in rspec regardless of ruby version. Both on ruby-2.6 and ruby-2.7 there are "6 errors occurred outside of examples".

**Comment by mbj** [comment]
Thats really interesting. I've to look into it in detail than. Thanks for the reproduction. Is it a public repo?

For the moment please set a subject ignore to your case.

**Comment by mbj** [comment]
@pawelpacana BTW when I reproduce this, on 2.7 the killfork actually runs into a segfault. Smells like an instance of a specific ruby bug we had seen in the past.

**Comment by mbj** [comment]
@pawelpacana I'd love to reproduce your behavior, not my segfault. Can you give me a bit more information on your environment? Which OS are you using, can you reproduce this on CI?

**Comment by mbj** [comment]
I've created #1055 to improve reporting on this and similar issue. Its not yet a solution.

As ruby segfaults for me (and produces incorrect results for you) I've to assume its actually a ruby-2.7 bug.

**Comment by mbj** [comment]
@pawelpacana Aha, after reading your original report more carefully:

> `#<Process::Status: pid 78659 SIGILL (signal 4)>`

Caught my eye: This is "worse" than a segfault, an illegal instruction terminating the killfork without printing the typical ruby segfault handler messages.

So its a 2.7 specific ruby bug. It was not reproducible in hand mutating, likely because in that environment there is no fork involved / GC triggered in different sequence. Also timeouts from the `Timeout` library can cause this.

Conceptually mutant typically assumes MRI "does not crash", and any time MRI does crash mutant should assume something very bad happened not showing these crashes as "covered". But given the ongoing problems where MRI has these kinds of issues, I'm thinking about counting segfaults and illegal instructions as killed, but report them separately.

Do you concur here?

**Comment by pawelpacana** [comment]
@mbj

on macos:

```
  * mutant (0.9.12 d8b56bb)
	Summary: 
	Homepage: https://github.com/mbj/mutant
	Path: /Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f
  * mutant-rspec (0.9.12 d8b56bb)
	Summary: Rspec integration for mutant
	Homepage: https://github.com/mbj/mutant
	Path: /Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f

ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-darwin19]

Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64

Mutant environment:
Matcher:         #<Mutant::Matcher::Config match_expressions: [AggregateRoot.included]>
Integration:     rspec
Jobs:            8
Includes:        []
Requires:        ["aggregate_root"]
Subjects:        1
Total-Tests:     35
Selected-Tests:  31
Tests/Subject:   31.00 avg
Mutations:       14
AggregateRoot.included:/Users/pawelpacana/Code/rails_event_store/aggregate_root/lib/aggregate_root.rb:86
- rspec:0:./spec/aggregate_root_spec.rb:9/AggregateRoot should have ability to apply event on itself
- rspec:1:./spec/aggregate_root_spec.rb:19/AggregateRoot brand new aggregate does not have any unpublished events
- rspec:2:./spec/aggregate_root_spec.rb:24/AggregateRoot should receive a method call based on a default apply strategy
- rspec:3:./spec/aggregate_root_spec.rb:32/AggregateRoot should raise error for missing apply method based on a default apply strategy
- rspec:4:./spec/aggregate_root_spec.rb:38/AggregateRoot should ignore missing apply method based on a default non-strict apply strategy
- rspec:5:./spec/aggregate_root_spec.rb:44/AggregateRoot should receive a method call based on a custom strategy
- rspec:6:./spec/aggregate_root_spec.rb:52/AggregateRoot should return applied events
- rspec:7:./spec/aggregate_root_spec.rb:61/AggregateRoot should return only applied events
- rspec:8:./spec/aggregate_root_spec.rb:71/AggregateRoot #unpublished_events method is public
- rspec:9:./spec/aggregate_root_spec.rb:84/AggregateRoot #unpublished_events method does not allow modifying internal state directly
- rspec:10:./spec/aggregate_root_spec.rb:95/AggregateRoot.on generates private apply handler method
- rspec:11:./spec/aggregate_root_spec.rb:149/AggregateRoot.on handles super() with inheritance
- rspec:12:./spec/aggregate_root_spec.rb:186/AggregateRoot.on does not support anonymous events
- rspec:13:./spec/instrumented_repostory_spec.rb:9/AggregateRoot::InstrumentedRepository#load wraps around original implementation
- rspec:14:./spec/instrumented_repostory_spec.rb:18/AggregateRoot::InstrumentedRepository#load instruments
- rspec:15:./spec/instrumented_repostory_spec.rb:38/AggregateRoot::InstrumentedRepository#store wraps around original implementation
- rspec:16:./spec/instrumented_repostory_spec.rb:47/AggregateRoot::InstrumentedRepository#store instruments
- rspec:17:./spec/instrumented_repostory_spec.rb:72/AggregateRoot::InstrumentedRepository#with_aggregate instruments both load and store
- rspec:22:./spec/repository_spec.rb:20/AggregateRoot::Repository#initialize should use default client if event_store not provided
- rspec:23:./spec/repository_spec.rb:33/AggregateRoot::Repository#initialize should prefer provided event_store client
- rspec:24:./spec/repository_spec.rb:48/AggregateRoot::Repository#load 
- rspec:25:./spec/repository_spec.rb:55/AggregateRoot::Repository#load 
- rspec:26:./spec/repository_spec.rb:62/AggregateRoot::Repository#load 
- rspec:27:./spec/repository_spec.rb:70/AggregateRoot::Repository#load 
- rspec:28:./spec/repository_spec.rb:80/AggregateRoot::Repository#store 
- rspec:29:./spec/repository_spec.rb:95/AggregateRoot::Repository#store updates aggregate stream position and uses it in subsequent publish call as expected_version
- rspec:30:./spec/repository_spec.rb:120/AggregateRoot::Repository#with_aggregate 
- rspec:31:./spec/transform_spec.rb:7/AggregateRoot::Transform 
- rspec:32:./spec/transform_spec.rb:8/AggregateRoot::Transform 
- rspec:33:./spec/transform_spec.rb:9/AggregateRoot::Transform 
- rspec:34:./spec/transform_spec.rb:10/AggregateRoot::Transform 
evil:AggregateRoot.included:/Users/pawelpacana/Code/rails_event_store/aggregate_root/lib/aggregate_root.rb:86:d3c17
@@ -1,4 +1,4 @@
 def self.included(host_class)
-  host_class.include(with_default_apply_strategy)
+  host_class.include(self)
 end
-----------------------
Log messages (combined stderr and stdout):
[killfork] /Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/gems/rspec-mocks-3.9.1/lib/rspec/mocks/message_expectation.rb:101: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
[killfork] /Users/pawelpacana/Code/rails_event_store/ruby_event_store/lib/ruby_event_store/client.rb:27: warning: The called method `publish' is defined here
Killfork exited nonzero. Its result (if any) was ignored.
Process status:
#<Process::Status: pid 91643 SIGILL (signal 4)>
Killing the mutation resulted in an integration error.
This is the case when the tests selected for the current mutation
did not produce a test result, but instead an exception was raised.

This may point to the following problems:
* Bug in mutant
* Bug in the ruby interpreter
* Bug in your test suite
* Bug in your test suite under concurrency

The following exception was raised while reading the killfork result:

```
#<ArgumentError: marshal data too short>
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/isolation/fork.rb:98:in `load'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/isolation/fork.rb:98:in `read_child_result'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/isolation/fork.rb:69:in `call'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/gems/procto-0.0.3/lib/procto.rb:62:in `public_send'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/gems/procto-0.0.3/lib/procto.rb:62:in `block in initialize'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/isolation/fork.rb:169:in `block (2 levels) in call'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/isolation/fork.rb:32:in `block in with'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/isolation/fork.rb:31:in `pipe'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/isolation/fork.rb:31:in `with'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/isolation/fork.rb:168:in `block in call'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/isolation/fork.rb:32:in `block in with'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/isolation/fork.rb:31:in `pipe'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/isolation/fork.rb:31:in `with'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/isolation/fork.rb:167:in `call'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/env.rb:130:in `run_mutation_tests'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/env.rb:51:in `kill'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/parallel/worker.rb:28:in `call'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/parallel/worker.rb:28:in `block in call'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/parallel/worker.rb:23:in `loop'
/Users/pawelpacana/.rubies/ruby-2.7.1/lib/ruby/gems/2.7.0/bundler/gems/mutant-d8b56bb4b15f/lib/mutant/parallel/worker.rb:23:in `call'
```
-----------------------
Mutant environment:
Matcher:         #<Mutant::Matcher::Config match_expressions: [AggregateRoot.included]>
Integration:     rspec
Jobs:            8
Includes:        []
Requires:        ["aggregate_root"]
Subjects:        1
Total-Tests:     35
Selected-Tests:  31
Tests/Subject:   31.00 avg
Mutations:       14
Results:         14
Kills:           13
Alive:           1
Runtime:         0.26s
Killtime:        1.43s
Overhead:        -81.85%
Mutations/s:     54.13
Coverage:        92.86%
```

**Comment by pawelpacana** [comment]
I've made a brach where CI runs this mutation on 2.6 amd 2.7 on Linux and it segfaults on 2.7:https://github.com/RailsEventStore/rails_event_store/pull/767

**Comment by mbj** [comment]
@pawelpacana k, so its actually a ruby bug. But mutant should handle these a bit more gracefully.

I'd propose to change mutant to count these MRI crashes as "killed by ruby bug", and still report them separately.

**Comment by pawelpacana** [comment]
> I'd propose to change mutant to count these MRI crashes as "killed by ruby bug", and still report them separately.

Sounds like a pragmatic approach.


> Also timeouts from the Timeout library can cause this.

I've [excluded Timeout](https://github.com/RailsEventStore/rails_event_store/pull/767/commits/7cec5b1879ec4735236989400415b50c29539330) in reproducton branch.

**Comment by mbj** [comment]
I do not get why MRI does crash on these. anyhow threading this ones more gracefully is next on my list.

**Comment by mbj** [comment]
@pawelpacana Using:

```
coverage_criteria:
  process_abort: true
```

You can (once #1120) is merged and released with `0.10.8` instruct mutant to count all abnormal child process exits (this includes, segfauls, illegal instructions etc) as covered mutations.

Later, when mutant has subject specific config, I'll recommend to set this coverage criteria only on affected subjects. This will allow new cases that pop up to be noticed and trigger upstream reports of these problems. No pure ruby code (mutant is pure ruby, no C exts) should *ever* cause MRI to carsh on itself.
